api/src/main/java/org/apache/iceberg/types/CheckCompatibility.java

```diff
@@ -93,7 +93,7 @@ private CheckCompatibility(Schema schema, boolean checkOrdering, boolean checkNu
   }
 
   @Override
-  public List<String> struct(Types.StructType readStruct, Iterable<List<String>> fieldErrorLists) {
+  public ImmutableList<String> struct(Types.StructType readStruct, Iterable<List<String>> fieldErrorLists) {
```

Comment:
We don't typically use `ImmutableList` to avoid leaking it in methods that are accidentally public or purposely part of the API. I'm +1 for returning `ImmutableList.copyOf(errors)` below, but I don't think we should guarantee the list will be an `ImmutableList`.

Reply:
Then we should disable this warning:
```
/Users/fokkodriesprong/Desktop/incubator-iceberg/api/src/main/java/org/apache/iceberg/types/CheckCompatibility.java:96: warning: [MixedMutabilityReturnType] This method returns both mutable and immutable collections or maps from different paths. This may be confusing for users of the method.
  public List<String> struct(Types.StructType readStruct, Iterable<List<String>> fieldErrorLists) {
```
Personally I prefer to change the signature to `ImmutableList`.

Reply:
I think what that message is saying is that sometimes the method returns a mutable list and sometimes it returns an immutable one. I don't think the signature change is required to solve the problem. We just need to make sure that all paths return either immutable or mutable.

We can't change the signature because we don't want to leak Guava in the Iceberg API.